### PR TITLE
JWT-based authentication using local SQLite3 database

### PR DESCRIPTION
**Pull request by rcook**
Chris: Here is the local authentication implementation I outlined at BayHac last month. Without additional configuration, CodeWorld should continue to work exactly as it does now, using Google authentication as the standard authentication method. With some configuration, as described in `README.md`, the local authentication method (accounts stored in SQLite3 database and JWT-based authentication) will be enabled.

-Richard

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +132,73 @@ following resources:
...
+```
+
+##### Create a JWT secret
+
+To use local authentication, you will also need to generate a JWT secret. This
+is used to sign JWT tokens passed back and forth between the server and the
+browser. From the Git repository's root directory, run the following command:
+
+```
+build/bin/codeworld-auth generate-secret local-auth-secret.txt
```
[the remark this answers is not in the document]
Comment (rcook) [reply README.md]: `codeworld-account` manages the accounts database while `codeworld-auth` manages authentication. I realize that this is a distinction without a difference for most people, so I will merge the functionality into a single tool.

File: codeworld-account/codeworld-account.cabal [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+source-repository head
+  type:                                     git
+  location:                                 https://github.com/google/codeworld.git
+
+library
+  default-language:                         Haskell2010
+  hs-source-dirs:                           lib
+  ghc-options:                              -W
+                                            -Wall
+--                                            -Werror=incomplete-patterns
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-account/codeworld-account.cabal]: Fixed.

File: codeworld-account/app/Main.hs [hunk 3]
```diff
@@ -0,0 +1,216 @@
+{-
+  Copyright 2018 The CodeWorld Authors. All rights reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-account/app/Main.hs]: Fixed.

File: codeworld-account/lib/CodeWorld/Account/Actions.hs [hunk 4]
```diff
@@ -0,0 +1,137 @@
+{-
+  Copyright 2018 The CodeWorld Authors. All rights reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-account/lib/CodeWorld/Account/Actions.hs]: Fixed.

File: codeworld-account/lib/CodeWorld/Account.hs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+{-
+  Copyright 2018 The CodeWorld Authors. All rights reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-account/lib/CodeWorld/Account.hs]: Fixed.

File: codeworld-auth/lib/CodeWorld/Auth/GoogleAuth.hs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    parseJSON (Object v) = User <$> (UserId <$> v .: "user_id") <*> v .: "audience"
+    parseJSON _ = mzero
+
+configureAuth :: FilePath -> IO (Maybe AuthConfig)
+configureAuth appDir = do
+    let clientIdPath = appDir </> "web/clientId.txt"
+    clientIdPathExists <- doesFileExist clientIdPath
+    case clientIdPathExists of
+        False -> do
+            putStrLn $ "Client ID file not found at " ++ clientIdPath ++ ": skipping configuration of Google authentication"
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-auth/lib/CodeWorld/Auth/GoogleAuth.hs]: Fixed.

File: codeworld-auth/lib/CodeWorld/Auth/LocalAuth.hs [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+authRoutes :: AuthConfig -> [(ByteString, Snap ())]
+authRoutes authConfig =
+    [ ("signIn", method POST (signInHandler authConfig))
+    ]
+
+authenticated :: (UserId -> Snap ()) -> AuthConfig -> Snap ()
+authenticated handler authConfig = do
+    req <- getRequest
+    case getHeader "Authorization" req of
+        Nothing -> finishWith unauthorized401
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-auth/lib/CodeWorld/Auth/LocalAuth.hs]: I have fixed this to pass pack a `WWW-Authenticate` header with value `Bearer`.

File: codeworld-auth/lib/CodeWorld/Auth/LocalAuth.hs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+parseAuthHeader :: ByteString -> Maybe JSON
+parseAuthHeader bs =
+    case splitOn " " (Char8.unpack bs) of
+        [authType, t] -> if authType /= "Bearer"
+                            then Nothing
+                            else Just (Text.pack t)
+        _ -> Nothing
+
+addAuthHeader :: AuthConfig -> UserId -> Snap ()
```
[the remark this answers is not in the document]
Comment (rcook) [reply codeworld-auth/lib/CodeWorld/Auth/LocalAuth.hs]: I have seen `Authorization` in response headers in other applications and nothing I've read explicitly states that this is wrong. However, I will look at coming up with a different mechanism to pass back the updated token.

**Comment by cdsmith** [comment]
Thanks for the fixes.  I definitely don't want to block you for too long on changes, if the current code works correctly.  I was just trying to understand what was going on and match it to my understanding of HTTP authentication and related specs.  It may well be that it's my understanding that needs fixing, rather than the code!

**Comment by rcook** [comment]
I'm going to rework the code so it doesn't need the Authorization header in the response. I'm also going to implement a proper refresh token workflow. The code will be better for it! Thanks!

**Pull request closed without merging**